bookwyrm/models/move.py

```diff
@@ -0,0 +1,87 @@
+""" move an object including migrating a user account """
+from django.core.exceptions import PermissionDenied
+from django.db import models
+
+from bookwyrm import activitypub
+from .activitypub_mixin import ActivityMixin
+from .base_model import BookWyrmModel
+from . import fields
+
+
+class Move(ActivityMixin, BookWyrmModel):
+    """migrating an activitypub user account"""
+
+    user = fields.ForeignKey(
+        "User", on_delete=models.PROTECT, activitypub_field="actor"
+    )
+
+    object = fields.CharField(
+        max_length=255,
+        blank=False,
+        null=False,
+        activitypub_field="object",
+    )
+
+    origin = fields.CharField(
+        max_length=255,
+        blank=True,
+        null=True,
+        default="",
+        activitypub_field="origin",
+    )
+
+    activity_serializer = activitypub.Move
+
+
+class MoveUser(Move):
+    """migrating an activitypub user account"""
+
+    target = fields.ForeignKey(
+        "User",
+        on_delete=models.PROTECT,
+        related_name="move_target",
+        activitypub_field="target",
+    )
+
+    def save(self, *args, **kwargs):
+        """update user info and broadcast it"""
+
+        # only allow if the source is listed in the target's alsoKnownAs
+        if self.user in self.target.also_known_as.all():
+
+            self.user.also_known_as.add(self.target.id)
+            self.user.update_active_date()
+            self.user.moved_to = self.target.remote_id
+            self.user.save(update_fields=["moved_to"])
+
+            if self.user.local:
+                kwargs[
+                    "broadcast"
+                ] = True  # Only broadcast if we are initiating the Move
+
+            super().save(*args, **kwargs)
+
+            for follower in self.user.followers.all():
+                if follower.local:
+                    MoveUserNotification.objects.create(user=follower, target=self.user)
+
+        else:
+            raise PermissionDenied()
+
+
+class MoveUserNotification(models.Model):
```

Comment:
I'm not understanding what this model does that can't be handled in the notification model? It looks like saving here triggers creating a Notification, but does it serve an additional function?

Reply:
This is a good question. I think I didn't understand how the notifications work and copied this pattern from somewhere else...

So this should just be an ordinary notification with the `target` as `related_user` and type of `MOVE`?

Reply:
The way it's created [in the listener](https://github.com/bookwyrm-social/bookwyrm/pull/2970/files#diff-ac98bf3171c4c7025f6ea238b24ed7e1056f52f0f0f77709e21462ba6df215d4R345) is correct -- you can just call `Notification.notify(...)` instead of [creating the MoveUserNotification](https://github.com/bookwyrm-social/bookwyrm/pull/2970/files#diff-0b203a68eea4a30f195d7ccd947b1557775264146348ee1f00a13c9ac5eed255R66) object

Reply:
yeah honestly I have no idea why I did it this way, it's kind of obvious now that I just created a totally unnecessary step in the middle.